File: LENTABP10/src/test/java/com/lenta/lentabp10/testWriteoffTask_ClearTask.java
package com.lenta.lentabp10;

import com.lenta.lentabp10.models.memory.MemoryTaskExciseStampRepository;
import com.lenta.lentabp10.models.memory.MemoryTaskProductRepository;
import com.lenta.lentabp10.models.memory.MemoryTaskRepository;
import com.lenta.lentabp10.models.memory.MemoryTaskWriteOffReasonRepository;
import com.lenta.lentabp10.models.repositories.ITaskExciseStampRepository;
import com.lenta.lentabp10.models.repositories.ITaskProductRepository;
import com.lenta.lentabp10.models.repositories.ITaskRepository;
import com.lenta.lentabp10.models.repositories.ITaskWriteOffReasonRepository;
import com.lenta.lentabp10.models.task.TaskDescription;
import com.lenta.lentabp10.models.task.TaskExciseStamp;
import com.lenta.lentabp10.models.task.TaskType;
import com.lenta.lentabp10.models.task.WriteOffReason;
import com.lenta.lentabp10.models.task.WriteOffTask;
import com.lenta.shared.models.core.MatrixType;
import com.lenta.shared.models.core.ProductInfo;
import com.lenta.shared.models.core.ProductType;
import com.lenta.shared.models.core.Uom;

import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;

import static org.junit.Assert.assertEquals;

public class testWriteoffTask_ClearTask {
    TaskDescription taskDescription;
    ITaskProductRepository taskProductRepository = new MemoryTaskProductRepository();
    ITaskExciseStampRepository taskExciseStampRepository = new MemoryTaskExciseStampRepository();
    ITaskWriteOffReasonRepository taskWriteOfReasonRepository = new MemoryTaskWriteOffReasonRepository();
    ITaskRepository taskRepository = new MemoryTaskRepository(taskProductRepository, taskExciseStampRepository, taskWriteOfReasonRepository);
    WriteOffTask task;

    public void creatingObjectsForTest() {
        taskDescription = new TaskDescription(new TaskType("СГП", "nСГП"),
                "Списание от 04.06 10:23",
                "0002",
                new ArrayList<>(Arrays.asList("949ВД")),
                new ArrayList<>(Arrays.asList("N")),
                new ArrayList<>(Arrays.asList("2FER", "3ROH"))
        );

        task = new WriteOffTask(taskDescription, taskRepository);
    }

    @Test
    public void testClearTask() {
        creatingObjectsForTest();

        ProductInfo product1 = new ProductInfo("materialNumber1", "description", new Uom("ST", "шт"), ProductType.General,
                false, 1, MatrixType.Active, "materialType");

        ProductInfo product2 = new ProductInfo("materialNumber2", "description", new Uom("ST", "шт"), ProductType.NonExciseAlcohol,
                false, 1, MatrixType.Active, "materialType");

        ProductInfo product3 = new ProductInfo("materialNumber3", "description", new Uom("ST", "шт"), ProductType.ExciseAlcohol,
                false, 1, MatrixType.Active, "materialType");

        WriteOffReason reason1 = new WriteOffReason("01", "Срок годности");
        WriteOffReason reason2 = new WriteOffReason("02", "Срок негодности");

        TaskExciseStamp exciseStamp1 = new TaskExciseStamp("materialNumber3", "1", "материал набора", "Срок годности", false);
        TaskExciseStamp exciseStamp2 = new TaskExciseStamp("materialNumber3", "2", "материал набора", "Срок негодности", false);


        task = task.processGeneralProduct(product1)
                .add(reason1, 1)
                .apply();

        task = task.processNonExciseAlcoProduct(product2)
                .add(reason1, 1)
                .add(reason2, 2)
                .apply();

        task = task.processExciseAlcoProduct(product3)
                .add(reason1, 1, exciseStamp1)
                .add(reason1, 1, exciseStamp2)
                .apply();

        task.clearTask();

        assertEquals(0, task.getProcessedProducts().size());
        assertEquals(0, task.getTotalCountOfProduct(product1),0);
        assertEquals(0, task.getTotalCountOfProduct(product2),0);
        assertEquals(0, task.getTotalCountOfProduct(product3),0);
    }
}

File: writeoff_breakage/src/test/java/com/lenta/bp10/testWriteoffTask_ClearTask.kt
package com.lenta.bp10

import com.lenta.bp10.models.memory.MemoryTaskExciseStampRepository
import com.lenta.bp10.models.memory.MemoryTaskProductRepository
import com.lenta.bp10.models.memory.MemoryTaskRepository
import com.lenta.bp10.models.memory.MemoryTaskWriteOffReasonRepository
import com.lenta.bp10.models.repositories.ITaskExciseStampRepository
import com.lenta.bp10.models.repositories.ITaskProductRepository
import com.lenta.bp10.models.repositories.ITaskRepository
import com.lenta.bp10.models.repositories.ITaskWriteOffReasonRepository
import com.lenta.bp10.models.task.*
import com.lenta.shared.models.core.MatrixType
import com.lenta.shared.models.core.ProductInfo
import com.lenta.shared.models.core.ProductType
import com.lenta.shared.models.core.Uom
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

class testWriteoffTask_ClearTask {
    lateinit var taskDescription: TaskDescription
    var taskProductRepository: ITaskProductRepository = MemoryTaskProductRepository()
    var taskExciseStampRepository: ITaskExciseStampRepository = MemoryTaskExciseStampRepository()
    var taskWriteOfReasonRepository: ITaskWriteOffReasonRepository = MemoryTaskWriteOffReasonRepository()
    var taskRepository: ITaskRepository = MemoryTaskRepository(taskProductRepository, taskExciseStampRepository, taskWriteOfReasonRepository)
    lateinit var task: WriteOffTask

    fun creatingObjectsForTest() {
        taskDescription = TaskDescription(
                TaskType("СГП", "nСГП"),
                "Списание от 04.06 10:23",
                "0002",
                ArrayList(Arrays.asList("949ВД")),
                ArrayList(Arrays.asList("N")),
                ArrayList(Arrays.asList("2FER", "3ROH")), "perNo", "printer", "tkNumber", "ipAdress"
        )

        task = WriteOffTask(taskDescription, taskRepository)
    }

    @Test
    fun testClearTask() {
        creatingObjectsForTest()

        val product1 = ProductInfo("materialNumber1", "description", Uom("ST", "шт"), ProductType.General,
                false, 1, MatrixType.Active, "materialType")

        val product2 = ProductInfo("materialNumber2", "description", Uom("ST", "шт"), ProductType.NonExciseAlcohol,
                false, 1, MatrixType.Active, "materialType")

        val product3 = ProductInfo("materialNumber3", "description", Uom("ST", "шт"), ProductType.ExciseAlcohol,
                false, 1, MatrixType.Active, "materialType")

        val reason1 = WriteOffReason("01", "Срок годности")
        val reason2 = WriteOffReason("02", "Срок негодности")

        val exciseStamp1 = TaskExciseStamp("materialNumber3", "1", "материал набора", "Срок годности", false)
        val exciseStamp2 = TaskExciseStamp("materialNumber3", "2", "материал набора", "Срок негодности", false)


        task = task.processGeneralProduct(product1)!!
                .add(reason1, 1.0)
                .apply()

        task = task.processNonExciseAlcoProduct(product2)!!
                .add(reason1, 1.0)
                .add(reason2, 2.0)
                .apply()

        task = task.processExciseAlcoProduct(product3)!!
                .add(reason1, 1.0, exciseStamp1)
                .add(reason1, 1.0, exciseStamp2)
                .apply()

        task.clearTask()

        assertEquals(0, task.getProcessedProducts().size.toLong())
        assertEquals(0.0, task.getTotalCountOfProduct(product1), 0.0)
        assertEquals(0.0, task.getTotalCountOfProduct(product2), 0.0)
        assertEquals(0.0, task.getTotalCountOfProduct(product3), 0.0)
    }
}